pipe/section/section_impls/mycelial_server/src/source.rs

```diff
@@ -149,46 +150,125 @@ impl Mycelial {
 
         let origin = match res.headers().get("x-message-origin") {
             None => Err("response needs to have x-message-origin header")?,
-            Some(v) => v.to_str().unwrap().to_string(),
+            Some(v) => v
+                .to_str()
+                .expect("bad header 'x-message-origin' value")
+                .to_string(),
+        };
+
+        let stream_type = match res.headers().get("x-stream-type") {
+            None => Err("response needs to have x-stream-type header")?,
+            Some(v) => v
+                .to_str()
+                .expect("bad header 'x-stream-type' value")
+                .to_string(),
         };
 
         let maybe_new_offset = match res.headers().get("x-message-id") {
             None => Err("response needs to have x-message-id header")?,
             // FIXME: unwrap
-            Some(v) => v.to_str().unwrap().parse().unwrap(),
+            Some(v) => v
+                .to_str()
+                .expect("bad header 'x-message-id' value")
+                .parse()
+                .unwrap(),
         };
 
         if maybe_new_offset == *offset {
             return Ok(None);
         }
         *offset = maybe_new_offset;
-
-        let body = res.bytes().await?.to_vec();
-        let len = body.len() as u64;
-        let mut body = Cursor::new(body);
-        let mut batches = vec![];
-        while body.position() < len {
-            let reader = StreamReader::try_new_unbuffered(&mut body, None).unwrap();
-            for batch in reader {
-                let batch = batch?;
-                batches.push(Some(batch.into()))
+        let o = *offset;
+        let weak_chan = section_chan.weak_chan();
+        match stream_type.as_str() {
+            "binary" => {
+                let stream = res.bytes_stream().map(|chunk| {
```

Comment:
Since this function is big I had some trouble with the smaller variable names, it might be easier to read either if some of this is extracted into another function(s), or variables like `res` were more descriptive.